### DISCUSSION TITLE
Using "convert" in user-defined functions in gen_ens_prod

### DISCUSSION THREAD

**Discussion by Jeff-Duda** [dtcenter/METplus#1948] [Configuration] [answered]
This discussion is intended as a bump to the discussion closed here: https://github.com/dtcenter/METplus/discussions/1287

I had a very similar question as the poster of that discussion, except I do not need to reference the climatological distribution.

Instead, I am querying the syntax of using the _convert_ function to derive a field in the config file from fields available in the input files.

In particular, I have 3D simulated reflectivity and I want to calculate the vertical max reflectivity using the configuration file options. I see the _convert(???)=max(???)_ availability, but what am I supposed to put in for the ???s? Also, what is the proper syntax for this in the config file? It appears that it belongs somewhere in the fcst.field dictionary entry, but what punctuation do I need to use?

I see the example in the User's Guide (https://met.readthedocs.io/en/latest/Users_Guide/config_options.html#fcst) that is

`convert(t) = max(1, sqrt(abs(t)));`

But this example appears to take the max between a 1D field and a fixed constant. I don't want to do that. Instead, I want the column maximum (vertical maximum) among one of the dimensions of the 3D input array. How do I specify such information in the config file?

**Comment by JohnHalleyGotway** [comment, marked as answer]
@Jeff-Duda, it sounds to me like you want to compute the maximum across multiple vertical slices of data. The `convert(x)` function will not be useful in this context. It is only applied to a single 2D plane of data... or point observations values... at once.

For example, the following 2 calls to `plot_data_plane` use convert to change units from Kelvin to Celsius for 2 different levels of data:
```
plot_data_plane sample.grib2 TMP_P500.ps 'name="TMP"; level="P500"; convert(x) = x - 273.15;'
plot_data_plane sample.grib2 TMP_P850.ps 'name="TMP"; level="P850"; convert(x) = x - 273.15;'
```
However, there is no way to call a `convert(x)` command to compute the maximum of the 500mb and 850mb temperature values.

Instead, you could using the `-derive` option in `pcp_combine`:
```
pcp_combine -derive max \
sample.grib 'name="TMP"; level="P500";' \
sample.grib 'name="TMP"; level="P850";' \
max_tmp.nc -name TMP_P500_P850_max
```
With this call, `pcp_combine` reads the two fields listed from the two input files listed, computes their maximum value, and writes the results to `max_tmp.nc` using the variable name `TMP_P500_P850_max`.

Perhaps that approach would work for you?

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion to see if questions remain. If not, please feel free to select the "Mark As Answer" button for one of the previous responses, and then I'll lock it to prevent future posts. If I don't hear otherwise in the next week or so, I'll just go ahead and mark it as being answered.

Thanks!

**Reply by Jeff-Duda** [reply]
Well...while I'm dissatisfied with the lack of capability of this tool to handle this operation, I suppose the answer provided is sufficient as an "answer" for the discussion, so I will mark it as such and you can lock it.

**Reply by JohnHalleyGotway** [reply]
Jeff, sounds good. So, in general, is it the case that you'd like to configure MET to define a range of levels and an operation to be preformed on each column of values? Perhaps configure it something like this:
```
name = "TMP";
level = "P500-1000";
derive = "max";
```
So perhaps we could introduce a `derive` config option to be consist with the `-derive` command line option of pcp_combine? So multiple "levels" of the same variable would be collapsed down into a single field. And we'd want to support the same derive options of pcp_combine (sum, min, max, range, mean, stdev, vld_count). So this would be like running pcp_combine first without actually having to do so. Perhaps, by convention, we'd append the derive option to the output variable name (i.e. "TMP_max") but of course let the user override that using the set_attr_name option (i.e. `set_attr_name = "MAXTMP";`).

Is that the sort of functionality you're looking for?

**Reply by Jeff-Duda** [reply]
Yes!

**Reply by JohnHalleyGotway** [reply]
Thanks for confirming. Please see GitHub issue dtcenter/MET#2359 to describe this feature request. I'll go ahead and lock conversation on this discussion. Please feel free to comment directly on dtcenter/MET#2359 with additional input.
